File: poker-voice/scripts/video-ocr-helper.py
#!/usr/bin/env python3
import json
import math
import sys


def emit(payload):
    sys.stdout.write(json.dumps(payload, ensure_ascii=False) + "\n")
    sys.stdout.flush()


def box_center(box):
    try:
        xs = [float(p[0]) for p in box]
        ys = [float(p[1]) for p in box]
        if not xs or not ys:
            return None, None
        return sum(xs) / len(xs), sum(ys) / len(ys)
    except Exception:
        return None, None


def main():
    if len(sys.argv) < 2:
        emit({"type": "error", "stage": "args", "message": "Usage: video-ocr-helper.py <videoPath> [sampleMs] [maxFrames] [startMs] [endMs]"})
        return 1

    video_path = sys.argv[1]
    sample_ms = max(100, int(sys.argv[2])) if len(sys.argv) >= 3 else 1000
    max_frames = max(1, int(sys.argv[3])) if len(sys.argv) >= 4 else 600
    start_ms = 0.0
    end_ms = None
    if len(sys.argv) >= 5:
        start_ms = max(0.0, float(sys.argv[4]))
    if len(sys.argv) >= 6:
        candidate_end = float(sys.argv[5])
        if math.isfinite(candidate_end) and candidate_end >= 0.0:
            end_ms = candidate_end
    if end_ms is not None and end_ms < start_ms:
        end_ms = start_ms

    try:
        import cv2
        from rapidocr_onnxruntime import RapidOCR
    except Exception as exc:
        emit({
            "type": "error",
            "stage": "imports",
            "message": f"Missing OCR dependencies: {exc}",
            "hint": "Install with: python3 -m pip install --target <path> rapidocr-onnxruntime"
        })
        return 2

    cap = cv2.VideoCapture(video_path)
    if not cap.isOpened():
        emit({"type": "error", "stage": "reader_open", "message": "Cannot open video file."})
        return 3

    fps = cap.get(cv2.CAP_PROP_FPS) or 0.0
    frame_count = cap.get(cv2.CAP_PROP_FRAME_COUNT) or 0.0
    width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH) or 0)
    height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT) or 0)
    duration_ms = (frame_count / fps * 1000.0) if fps > 0 and frame_count > 0 else 0.0

    emit({
        "type": "meta",
        "duration_ms": duration_ms,
        "width": width,
        "height": height,
        "fps": fps,
        "sample_ms": sample_ms,
        "max_frames": max_frames,
        "start_ms": start_ms,
        "end_ms": end_ms,
        "decoder": "opencv"
    })

    try:
        ocr = RapidOCR()
    except Exception as exc:
        cap.release()
        emit({"type": "error", "stage": "ocr_init", "message": str(exc)})
        return 4

    sample_interval_frames = 1
    if fps > 0:
        sample_interval_frames = max(1, int(round((sample_ms / 1000.0) * fps)))

    sampled = 0
    frame_index = int(cap.get(cv2.CAP_PROP_POS_FRAMES) or 0)
    read_index = 0
    if start_ms > 0.0:
        cap.set(cv2.CAP_PROP_POS_MSEC, start_ms)
        frame_index = int(cap.get(cv2.CAP_PROP_POS_FRAMES) or 0)

    try:
        while sampled < max_frames:
            ok, frame = cap.read()
            if not ok:
                break

            frame_index = int(cap.get(cv2.CAP_PROP_POS_FRAMES) or (frame_index + 1)) - 1

            if read_index % sample_interval_frames != 0:
                read_index += 1
                continue

            frame_ms = cap.get(cv2.CAP_PROP_POS_MSEC)
            if not isinstance(frame_ms, (int, float)) or not math.isfinite(frame_ms) or frame_ms < 0:
                frame_ms = (frame_index / fps * 1000.0) if fps > 0 else float(frame_index * sample_ms)

            if frame_ms < (start_ms - max(5.0, sample_ms * 0.2)):
                read_index += 1
                continue

            if end_ms is not None and frame_ms > (end_ms + max(5.0, sample_ms * 0.4)):
                break

            lines = []
            try:
                result, _ = ocr(frame)
                if result:
                    for item in result:
                        text = str(item[1]).strip() if len(item) > 1 else ""
                        conf = float(item[2]) if len(item) > 2 else 0.0
                        if not text or conf < 0.25:
                            continue

                        bbox = item[0] if len(item) > 0 else None
                        cx, cy = box_center(bbox) if bbox else (None, None)

                        lines.append({
                            "text": text,
                            "confidence": conf,
                            "bbox": bbox,
                            "cx": cx,
                            "cy": cy
                        })
            except Exception as exc:
                emit({"type": "warn", "stage": "ocr", "ms": frame_ms, "message": str(exc)})

            emit({
                "type": "frame",
                "ms": frame_ms,
                "frame_index": frame_index,
                "lines": lines,
                "observation_count": len(lines)
            })

            sampled += 1
            read_index += 1
    finally:
        cap.release()

    emit({"type": "done", "sampled": sampled, "reader_status": 1})
    return 0


if __name__ == '__main__':
    raise SystemExit(main())

File: poker-voice/scripts/video-ocr-helper.m
#import <Foundation/Foundation.h>
#import <AVFoundation/AVFoundation.h>
#import <Vision/Vision.h>
#import <CoreImage/CoreImage.h>
#import <CoreMedia/CoreMedia.h>
#import <CoreVideo/CoreVideo.h>

static NSString *fourCCToString(FourCharCode code) {
  char c[5];
  c[0] = (char)((code >> 24) & 0xFF);
  c[1] = (char)((code >> 16) & 0xFF);
  c[2] = (char)((code >> 8) & 0xFF);
  c[3] = (char)(code & 0xFF);
  c[4] = '\0';
  return [NSString stringWithCString:c encoding:NSMacOSRomanStringEncoding] ?: @"????";
}

static void emitJson(NSDictionary *payload) {
  NSError *error = nil;
  NSData *data = [NSJSONSerialization dataWithJSONObject:payload options:0 error:&error];
  if (!data) {
    fprintf(stderr, "json_encode_error:%s\n", [[error localizedDescription] UTF8String]);
    return;
  }
  fwrite(data.bytes, 1, data.length, stdout);
  fwrite("\n", 1, 1, stdout);
  fflush(stdout);
}

int main(int argc, const char * argv[]) {
  @autoreleasepool {
    if (argc < 2) {
      emitJson(@{ @"type": @"error", @"stage": @"args", @"message": @"Usage: video-ocr-helper <videoPath> [sampleMs] [maxFrames]" });
      return 1;
    }

    NSString *videoPath = [NSString stringWithUTF8String:argv[1]];
    NSInteger sampleMs = (argc >= 3) ? MAX(100, atoi(argv[2])) : 1200;
    NSInteger maxFrames = (argc >= 4) ? MAX(1, atoi(argv[3])) : 600;

    NSURL *url = [NSURL fileURLWithPath:videoPath];
    AVAsset *asset = [AVAsset assetWithURL:url];
    NSArray<AVAssetTrack *> *tracks = [asset tracksWithMediaType:AVMediaTypeVideo];
    if (tracks.count == 0) {
      emitJson(@{ @"type": @"error", @"stage": @"tracks", @"message": @"No video tracks found." });
      return 2;
    }

    AVAssetTrack *track = tracks.firstObject;
    NSString *codec = @"unknown";
    if (track.formatDescriptions.count > 0) {
      CMFormatDescriptionRef fd = (__bridge CMFormatDescriptionRef)track.formatDescriptions[0];
      codec = fourCCToString(CMFormatDescriptionGetMediaSubType(fd));
    }

    Float64 durationSec = CMTimeGetSeconds(asset.duration);
    emitJson(@{
      @"type": @"meta",
      @"duration_ms": @(isfinite(durationSec) && durationSec > 0 ? durationSec * 1000.0 : 0),
      @"width": @(track.naturalSize.width),
      @"height": @(track.naturalSize.height),
      @"codec": codec,
      @"sample_ms": @(sampleMs),
      @"max_frames": @(maxFrames)
    });

    NSError *readerError = nil;
    AVAssetReader *reader = [[AVAssetReader alloc] initWithAsset:asset error:&readerError];
    if (!reader) {
      emitJson(@{ @"type": @"error", @"stage": @"reader_init", @"message": readerError.localizedDescription ?: @"Unable to initialize AVAssetReader." });
      return 3;
    }

    NSDictionary *settings = @{ (id)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_32BGRA) };
    AVAssetReaderTrackOutput *output = [[AVAssetReaderTrackOutput alloc] initWithTrack:track outputSettings:settings];
    output.alwaysCopiesSampleData = NO;
    if (![reader canAddOutput:output]) {
      emitJson(@{ @"type": @"error", @"stage": @"reader_output", @"message": @"Cannot add video output to AVAssetReader." });
      return 4;
    }
    [reader addOutput:output];

    if (![reader startReading]) {
      emitJson(@{ @"type": @"error", @"stage": @"reader_start", @"message": reader.error.localizedDescription ?: @"Cannot decode video stream." });
      return 5;
    }

    CIContext *ciContext = [CIContext contextWithOptions:nil];
    double lastSampledMs = -1;
    NSInteger sampled = 0;

    while (sampled < maxFrames) {
      CMSampleBufferRef sample = [output copyNextSampleBuffer];
      if (!sample) {
        break;
      }

      CMTime pts = CMSampleBufferGetPresentationTimeStamp(sample);
      Float64 seconds = CMTimeGetSeconds(pts);
      if (!isfinite(seconds) || seconds < 0) {
        CFRelease(sample);
        continue;
      }

      double frameMs = seconds * 1000.0;
      if (lastSampledMs >= 0 && (frameMs - lastSampledMs) < sampleMs) {
        CFRelease(sample);
        continue;
      }

      CVImageBufferRef imageBuffer = CMSampleBufferGetImageBuffer(sample);
      if (!imageBuffer) {
        CFRelease(sample);
        continue;
      }

      CIImage *ciImage = [CIImage imageWithCVPixelBuffer:imageBuffer];
      CGRect extent = ciImage.extent;
      CGImageRef cgImage = [ciContext createCGImage:ciImage fromRect:extent];
      if (!cgImage) {
        CFRelease(sample);
        continue;
      }

      VNRecognizeTextRequest *request = [[VNRecognizeTextRequest alloc] init];
      request.recognitionLevel = VNRequestTextRecognitionLevelFast;
      request.usesLanguageCorrection = NO;
      request.recognitionLanguages = @[ @"en-US" ];

      VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCGImage:cgImage options:@{}];
      NSError *ocrError = nil;
      BOOL ok = [handler performRequests:@[ request ] error:&ocrError];
      CGImageRelease(cgImage);

      if (!ok) {
        emitJson(@{ @"type": @"warn", @"stage": @"ocr", @"ms": @(frameMs), @"message": ocrError.localizedDescription ?: @"OCR failed on frame." });
        CFRelease(sample);
        lastSampledMs = frameMs;
        sampled += 1;
        continue;
      }

      NSMutableArray *lines = [NSMutableArray array];
      NSArray<VNRecognizedTextObservation *> *results = request.results ?: @[];
      for (VNRecognizedTextObservation *observation in results) {
        VNRecognizedText *top = [[observation topCandidates:1] firstObject];
        if (!top) continue;
        NSString *text = [[top.string stringByReplacingOccurrencesOfString:@"\n" withString:@" "] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
        if (text.length == 0) continue;
        if (top.confidence < 0.15f) continue;
        [lines addObject:@{ @"text": text, @"confidence": @(top.confidence) }];
      }

      emitJson(@{
        @"type": @"frame",
        @"ms": @(frameMs),
        @"lines": lines,
        @"observation_count": @(results.count)
      });

      CFRelease(sample);
      lastSampledMs = frameMs;
      sampled += 1;
    }

    if (reader.status == AVAssetReaderStatusFailed) {
      emitJson(@{ @"type": @"error", @"stage": @"reader_runtime", @"message": reader.error.localizedDescription ?: @"Reader failed while sampling." });
      return 6;
    }

    emitJson(@{ @"type": @"done", @"sampled": @(sampled), @"reader_status": @(reader.status) });
  }
  return 0;
}
